### PR TITLE
Fix nest and add test

### PR DESCRIPTION
**Pull request by wsmoses**

### CODE REVIEW

**Comment by vchuravy** [comment]
`git add -p` is great to not check-in Project.toml changes

File: Project.toml [hunk 1]
```diff
@@ -18,7 +18,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 [compat]
 Adapt = "3.3"
 CEnum = "0.4"
-Enzyme_jll = "0.0.40"
+Enzyme_jll = "0.0.39"
```
Comment (vchuravy) [inline Project.toml:21]: Was this a test? Or a intentional downgrade?

File: src/Enzyme.jl [hunk 2]
```diff
@@ -394,7 +402,7 @@ f(x) = x*x
 """
 @inline function autodiff(::ForwardMode, f::F, ::Type{A}, args...) where {F, A<:Annotation}
     args′  = annotate(args...)
-    if any_active(args′...)
+    if any_active(args...)
```
Comment (vchuravy) [inline src/Enzyme.jl:405]: Why this change?

File: src/compiler.jl [hunk 3]
```diff
@@ -4610,13 +4619,26 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
...
+        tape = API.EnzymeExtractTapeTypeFromAugmentation(augmented)
+        utape = API.EnzymeExtractUnderlyingTapeTypeFromAugmentation(augmented)
+        if utape != C_NULL
+            TapeType = EnzymeTapeToLoad{LLVMType(tape_type(utape))}
+        elseif tape != C_NULL
+            TapeType = tape_type(LLVMType(tape))
+        else
+            TapeType = Cvoid
+        end
+        @safe_show TapeType, tape, utape
```
Comment (vchuravy) [inline src/compiler.jl:4631]:
```suggestion
```

File: src/compiler.jl [hunk 4]
```diff
@@ -4639,12 +4661,30 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
...
+    jltype = convert(LLVM.LLVMType, combinedReturn; ctx)
 
-    # sret argument
-    if !isempty(T_JuliaSRet)
-        ST = LLVM.StructType(T_JuliaSRet; ctx)
-        pushfirst!(T_wrapperargs, LLVM.PointerType(ST, 10))
+    returnRoots = false
+    if uses_sret
+    	returnRoots = deserves_rooting(jltype)
+		if returnRoots
```
Comment (vchuravy) [inline src/compiler.jl:4683]: Please cleanup formatting

File: src/compiler.jl [hunk 5]
```diff
@@ -4781,11 +4829,52 @@ function create_abi_wrapper(enzymefn::LLVM.Function, F, argtypes, rettype, actua
...
+        if T_ret != T_void
+            ret!(builder, load!(builder, sret))
+        else
+            ret!(builder)
+        end
     end
 
     # make sure that arguments are rooted if necessary
     reinsert_gcmarker!(llvm_f)
+    @safe_show llvm_f
```
Comment (vchuravy) [inline src/compiler.jl:4877]:
```suggestion
```

File: src/compiler.jl [hunk 6]
```diff
@@ -5856,27 +5945,45 @@ end
...
+
+
+    combinedReturn = AnonymousStruct(Tuple{sret_types...})
+    uses_sret = is_sret(combinedReturn, ctx)
+    jltype = convert(LLVM.LLVMType, combinedReturn; ctx)
+    
+
+    returnRoots = false
+    if uses_sret
+    	returnRoots = deserves_rooting(jltype)
```
Comment (vchuravy) [inline src/compiler.jl:5959]: Please cleanup formatting

File: Project.toml (hunk 1 above)
Comment (wsmoses) [reply Project.toml:21]: Didnt intend to add Project.toml, just locally hadn't propagated and was lazy.

File: src/Enzyme.jl (hunk 2 above)
Comment (wsmoses) [reply src/Enzyme.jl:405]: any_active wasn't getting optimized out (which presumably was fixed by the tuple defn rather than recurrence, rather than this).

**Review by vchuravy — COMMENTED** [review]
Fix alloc_obj type for 1.6/1.7
